### PR TITLE
#8 implement login function

### LINKED ISSUE #8: Implement login function
## Summary 

#1 `POST /login`

This task is the implementation of the user login function.
This function returns JSON Web Token(JWT) with 200 HTTP status code, if username and password is valid.

### PR DESCRIPTION
**Pull request by yamakenji24**
Implement /login endpoint which authorize user and return jwt. 
Secret key used to generate jwt is in instance, and must be hidden later. 
Faked user information is used, and must be changed to reading from the mysql data. Done after #25 is merged.

### CODE REVIEW

File: auth.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+}
+# for instance 
+secret_key = '09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7'
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/login")
+
+def create_tokens(user_id: str):
+  payload = {
+    'token_type': 'access_token',
+    'exp': datetime.utcnow() + timedelta(minutes=2),
```
Comment (wakatakeru) [inline auth.py]: [REQUEST CHANGES]

Expiration date definitions would not better hard-coded. For example, there is a method to read the period from the environment variable.

Comment (wakatakeru) [reply auth.py:24]: May also need to define a time zone.

File: auth.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      "id": "base64id",
+      "username": "yamakenji",
+      "full_name": "Kenji Yamashita",
+      "email": "sample@example.com",
+      "password": "fakehashedsecret",
+      "disabled": False,
+  }
+}
+# for instance 
+secret_key = '09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7'
```
Comment (wakatakeru) [inline auth.py]: confirmed that using fake information for the time being.

File: auth.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/login")
+
+def create_tokens(user_id: str):
+  payload = {
+    'token_type': 'access_token',
+    'exp': datetime.utcnow() + timedelta(minutes=2),
+    'sub': user_id,
+    'iss': 'change it using env later'
```
Comment (wakatakeru) [inline auth.py:31]: confirmed that using fake information for the time being.

File: auth.py (hunk 2 above)
Comment (wakatakeru) [reply auth.py]: [REQUEST CHANGES]

I think it is necessary to change the method of reading **the secret key** from the file.
It is also better to read the location of **the secret key file** from the environment variable.

File: auth.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  return jwt.encode(payload, secret_key, algorithm='HS256')
+
+def authenticate(username: str, password: str):
+  user = fake_users_db.get(username)
```
Comment (wakatakeru) [inline auth.py:37]: confirmed that using fake information for the time being.

File: auth.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  if not user:
+    raise HTTPException(status_code=401, detail="Incorrect username or password")
+
+  if user["password"] != password:
```
Comment (wakatakeru) [inline auth.py:41]: Let's actually collate with a password hash. (I understand that it will change in the future.)

File: docker-compose.dev.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    depends_on:
+      - mysql
+    ports:
+      - "3000:3000"
+  mysql:
+    image: mysql
+    environment:
+      MYSQL_ROOT_PASSWORD: root
+    volumes:
+      - mysql-vol
```
Comment (wakatakeru) [inline docker-compose.dev.yml]: `mysql-vol` would be better `mysql-data`. In addition, it's a good idea to mount it to `/var/lib/mysql`.

File: docker-compose.dev.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+volumes:
+  mysql-vol:
```
Comment (wakatakeru) [inline docker-compose.dev.yml]: `mysql-vol` would be better `mysql-data`.

**Review by wakatakeru — CHANGES_REQUESTED** [review]
Request changes regarding constant management.

File: docker-compose.dev.yml (hunk 6 above)
Comment (wakatakeru) [reply docker-compose.dev.yml]: Please refer to `docker-compose.yml` of main branch.

File: docker-compose.dev.yml (hunk 7 above)
Comment (yamakenji24) [reply docker-compose.dev.yml]: thx. changing as mentioned.

File: docker-compose.dev.yml (hunk 6 above)
Comment (yamakenji24) [reply docker-compose.dev.yml]: Thx. Changed as mentioned.

File: auth.py (hunk 5 above)
Comment (yamakenji24) [reply auth.py:41]: Yep. Going to change in the future.

File: auth.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    raise HTTPException(status_code=401, detail="Incorrect username or password")
+  
+  return {
+    "access_token": create_tokens(user["id"]),
+    "token_type": "bearer"
+  }
+
```
Comment (wakatakeru) [inline auth.py]: unnecessary spaces.

File: auth.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      "email": "sample@example.com",
+      "password": "fakehashedsecret",
+      "disabled": False,
+  }
+}
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/login")
+
+exp_duration = os.getenv("JWT_EXP_SECONDS", 120)
+private_key_path = os.getenv("PRIVATE_KEY_PATH", "./key/private-key.pem")
```
Comment (wakatakeru) [inline auth.py:21]: Great.

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1 @@
+.key/*.pem
```
Comment (wakatakeru) [inline .gitignore]: please change directory to `key` from `.key`

Comment (wakatakeru) [reply .gitignore]:
```
$ mv .key key
```

File: auth.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+with open(private_key_path, "rb") as key_file:
+    private_key = key_file.read()
+    
+def create_tokens(user_id: str):
+  payload = {
+    'token_type': 'access_token',
+    'exp': datetime.utcnow() + timedelta(seconds=exp_duration),
```
Comment (wakatakeru) [inline auth.py:29]: Seems good.

**Review by wakatakeru — CHANGES_REQUESTED** [review]
please move directory

**Review by wakatakeru — APPROVED** [review]
LGTM!

**Comment by yamakenji24** [comment]
Using RS256 to encode, decode in pyjwt requires cryptography. 
Requested in #26

**Pull request merged**